services/pipeline/src/main/java/tech/pegasys/pantheon/services/pipeline/AsyncOperationProcessor.java

```diff
@@ -69,26 +77,48 @@ private void outputNextCompletedTask(final WritePipe<O> outputPipe) {
     } catch (final InterruptedException e) {
       LOG.trace("Interrupted while waiting for processing to complete", e);
     } catch (final ExecutionException e) {
-      LOG.error("Processing failed and we don't handle exceptions properly yet", e);
+      throw new RuntimeException("Async operation failed", e);
     } catch (final TimeoutException e) {
       // Ignore and go back around the loop.
     }
   }
 
-  @SuppressWarnings("rawtypes")
   private void waitForAnyFutureToComplete()
       throws InterruptedException, ExecutionException, TimeoutException {
-    CompletableFuture.anyOf(inProgress.toArray(new CompletableFuture[0])).get(1, TimeUnit.SECONDS);
+    nextOutputAvailableFuture.get(1, TimeUnit.SECONDS);
   }
 
   private void outputCompletedTasks(final WritePipe<O> outputPipe) {
+    boolean inProgressChanged = false;
     for (final Iterator<CompletableFuture<O>> i = inProgress.iterator(); i.hasNext(); ) {
       final CompletableFuture<O> process = i.next();
       final O result = process.getNow(null);
       if (result != null) {
+        inProgressChanged = true;
         outputPipe.put(result);
         i.remove();
+      } else if (preserveOrder) {
+        break;
       }
     }
+    if (inProgressChanged) {
+      updateNextOutputAvailableFuture();
+    }
+  }
+
+  /**
+   * CompletableFuture.anyOf adds a completion handler to every future its passed so if we call it
```

Comment:
Does it help to cancel the future returned from anyOf when you overwrite it??

Reply:
Sadly no, it looks like it only cleans the completion stack when the underlying futures complete.